### ISSUE TITLE
Multifluid test case broken 

### ISSUE THREAD

**Issue by prudhomm** [feelpp/feelpp#1199]
In develop and master the multifluid test case fails in json parsing probably

**Comment by prudhomm** [comment]
@metivett It fails also in `feature/ls`.

**Comment by metivett** [comment]
In `develop`, the problem seems to come from `advectionbase.cpp:230`. More precisely, I think it is due to an unauthorized access to non-initialized `ModelModels` shared_ptr. I need to check with @vincentchabannes ...

**Comment by metivett** [comment]
However, I checked again the version in `feature/ls` with a completely fresh install, starting from scratch, and I confirm that the `drop/drop2d.cfg` testcase (I ran exactly `mpirun -np 4 feelpp_toolbox_multifluid_2d --config-files drop/drop2d.cfg`) does work as expected. Could you please provide more details about your failure in `feature/ls`, @prudhomm ? It is weird...

**Comment by prudhomm** [comment]
@metivett thanks for checking this! it was @raouajaraib who tested the `feature/ls`
see [here](https://feelpp.slack.com/archives/CB2GDR9NE/p1531324050000280) however I haven't checked with her if she did it properly. @raouajaraib  could you explain what you did to get the `feature/ls` version? @metivett does not have the issues you have to run the testcase

**Comment by vincentchabannes** [comment]
@metivett @raouajaraib I understand the problem, the advection model with levelset  prefix has no json, and consequently the modelproperties is not initialized (modified since few weeks/month). In waiting for the fix, you can add  `--multifluid.levelset.filename=bid` and the modelproperties is initialized. Also there is no `Materials` section in the fluid json, it's required now. For example, for drop2d case, we need to add : 
```
    "Materials":
    {
        "OmegaFluid":
        {
            "rho":1000,
            "mu":10
        }
    },
```

**Comment by raouajaraib** [comment]
@prudhomm I changed the branch develop to feature/ls and I made a copy of the data files that I found there in my computer and I run drop2d, using this command feelpp_toolbox_multifluid_2d --config-file /feel/multifluid/feature/drop2d.cfg and I had the same error.

**Comment by prudhomm** [comment]
Did you recompile before running the app?

**Comment by raouajaraib** [comment]
No I didn't, that was the problem

**Comment by prudhomm** [comment]
@metivett when do you think you can fix this?

**Comment by metivett** [comment]
Hi @vincentchabannes ! I merge `develop` in `feature/ls`, and after the resolution of some conflicts, I am a bit stuck regarding a change you introduced with the management of `ModelProperties`. More specifically, the issue comes from the conditional allocation of `M_modelProps` in `toolboxes/feel/feelmodels/modelcore/modelnumerical.cpp:93`. This causes every model without a modelProperties `filename` to crash whenever accessing `modelProperties()`, and this is the case for many "sub`-Advection`s" within the `LevelSet` toolbox, which are automatically parametrised by the toolbox itself and do not require a real `filename`. 
I think the solution is either to provide an empty constructor for `ModelProperties` and use it as a default for `M_modelProps`, or test for the validity of the properties with `if( this->modelPropertiesPtr() )` whenever accessing it. The former seems more simple, but which solution do you prefer ?

**Comment by vincentchabannes** [comment]
@metivett We can't define a default constructor when filename is not given. In mutliphysics context (for example with thermoelectric), we need to check if the model property is given by the subphysics or by the mutliphysics toolboxes. You can write something like that : 
```
if ( !this->modelPropertiesPtr() )
        this->setModelProperties( std::make_shared<ModelProperties>( "", this->repository().expr(), this->worldComm() ) );
```
Probably, we can define in ModelNumerical a method() that init a empty ModelProperties.

Also,  it would be nice if you could follow the style / behavior of other toolboxes in your develop( constructor only init option parameters, the init() method do all the initialization (json,mesh,space,...))
This summer, I have tried to update the advection/levelset/multifluid but when I see all changes in your branches I stopped (I will send by slack the path of my git repo on atlas with theses changes, maybe it can help you).

**Comment by metivett** [comment]
@vincentchabannes Regarding the `ModelProperties`, maybe we could then just put something like you propose directly in `toolboxes/feel/feelmodels/modelcore/modelnumerical.cpp:93`, to get 
```
std::string modelPropFilename = Environment::expand( soption( _name=prefixvm(this->prefix(),"filename")) );
if ( !modelPropFilename.empty() )
    M_modelProps = std::make_shared<ModelProperties>( modelPropFilename, this->repository().expr(), this->worldComm() );
else
    M_modelProps = std::make_shared<ModelProperties>( "", this->repository().expr(), this->worldComm() ) );
```
Do you agree ?

**Comment by vincentchabannes** [comment]
@metivett No, as I say before, we can't initialize by default the modelprop. Some toolboxes ( in multiphysics context, the sub toolboxes) can have a model property which is initialized by another toolbox and I need to check if it's the case

**Comment by metivett** [comment]
@vincentchabannes ok, I patched only the `Advection` toolbox, and left `ModelProperties` unchanged.
Regarding the style of the toolboxes, I still have some reluctance dropping the `build` methods, since they are used as overloaded APIs to control the reuse of resources (e.g. I often have the `build(mesh_ptrtype)` and `build(space_ptrtype)` overloads to reuse an already constructed mesh or space. I think that transferring these overloads to `init(...)` would obfuscate the purpose of the `init` method.
What do you think ?

**Comment by vincentchabannes** [comment]
@metivett it's too complex maintain this kind of code and it's not scalable I think. What I do is : 
 1) build the toolbox object (init only worldComm and command line options)
 2) you can  initiliaze some data structure by calling for example setMesh or what you want (In basic toolbox, there is nothing todo)
3) call init() function: this initializes all datas that are not initialized

**Comment by metivett** [comment]
Ok, if you insist :wink: ! I will shortly push a working version of the merge of `develop` into `feature/ls`, and I start to work to standardise my toolboxes.

**Comment by metivett** [comment]
@vincentchabannes @prudhomm I have refactored the `Advection` class to remove `AdvectionBase` (and have only one `Advection` class), and homogenised the `init` process (no `build` method anymore). 
However, I am facing a small name conflict, and am thinking to resolve it by renaming the `Advection` class to something like `ADR`. Does this name suit you, or do you have any other proposition to better reflect what the toolbox actually solves for ?

**Comment by metivett** [comment]
Or maybe `AdvDiffReac` would be more explicit ?

**Comment by metivett** [comment]
@vincentchabannes Then I am a bit stuck... As I explained above, I need this for `Multifluid`, where the initialization of the algebraic structures requires *both*  the `Fluid` and the `Levelset` to be initialized before.

**Comment by metivett** [comment]
In particular when I want to impose an "inextensibility" condition on the interface

**Comment by vincentchabannes** [comment]
why not call levelset->init() before fluid->init()? levelset should not depend of fluid

**Comment by vincentchabannes** [comment]
look heatfluid or thermoectric toolbox. The goal is to have a coupling without (or few) intrusion in subtoolbox. It's really necessary to maintain and increase the complexity. Also, the blockvector is a small structure and you can build and upgrade/rebuild  after

**Comment by metivett** [comment]
At the moment, `Levelset` and `Fluid` are independent, but `MultiFluid` couples them. In particular, I reuse some spaces from the `Fluid` part in the `Levelset` toolbox to avoid having to many function spaces... 
For now, I had indeed chosen as a quick fix to first build the blockVector with an empty "subvector", and I rebuild it after both `FluidMechanics` and `Levelset` are properly initialized. But I find this approach not so clean...

**Comment by metivett** [comment]
@vincentchabannes  I have looked at the `HeatFluid` toolbox, and I understand your point. But I still think that the design of initializing the toolbox *and* at the same time the algebraic data is not well-suited for the coupling of toolboxes, where often the toolboxes (and their data) themselves are unrelated, but the corresponding multiphysics algebraic problem tightly couples both.
For the moment, as a quick fix, we could have another `bool` parameter in the `init` method (e.g. `initBlockVector`) to control the call to `this->buildBlockVector()` (similar to `buildAlgebraicFactory`). But this would only be a temporary solution; we should discuss more broadly this design...

**Comment by metivett** [comment]
This issue is solved in `feature/ls` (b3cd9d6e625e3221fc0c7f05985df9f2df6d6bdc)

**Comment by prudhomm** [comment]
I close this issue, it is in develop now
